src/Microsoft.AspNet.SignalR.JS/jquery.signalR.hubs.js

```diff
@@ -98,7 +98,9 @@
             /// <param name="callback" type="Function">The callback to be invoked.</param>
             /// <param name="callbackIdentity" type="Object">An optional object to use as the "identity" for the callback when checking if the handler has already been registered. Defaults to the value of 'callback' if not provided.</param>
             var that = this,
-                callbackMap = that._.callbackMap;
+                callbackMap = that._.callbackMap,
+                // https://github.com/SignalR/SignalR/issues/4310
+                isFromOldGeneratedHubProxy = !callbackIdentity && callback.toString().includes("// Call the client hub method");
```

Comment:
Yes, this is super gross. Generally, `hub.off` MUST be called with the same callback instance or callbackIdentity used during `hub.on`. The problem is that hub proxies generated by the SignalR server prior to 2.4.0 call `hub.on(eventName, callback)` and `hub.off(eventName, callback)` without a callbackIdentity and with *two different* callback instances.

We really should keep the old generated hub proxies working with the new client. The best way I've found so far to detect whether or not on/off is being called from the old generated hub proxy, is to look for the [// Call the client hub method](https://github.com/SignalR/SignalR/blob/d8f3843c17e382b238b64ade2548ead16f4d66d2/src/Microsoft.AspNet.SignalR.Core/Scripts/hubs.js#L24) comment from the generated function. If you know of any better alternatives, please tell me.

This wouldn't work if someone pre-generated the proxy using `signalr.exe ghp` and then minified it or something like that, but I'm not sure of any solution that handles that without completely reverting #4249.

This also means that there could be some issues if someone both uses the old generated hub proxy and manually calls `hub.on` for the same event using a function containing "// Call the client hub method", but I don't think that's a likely scenario that we should worry about.

Reply:
Ugh, server-owned client-side code is the W O R S T :)